mesonbuild/build.py

```diff
@@ -1969,6 +1969,11 @@ def __init__(
             environment: environment.Environment,
             compilers: T.Dict[str, 'Compiler'],
             kwargs):
+        # See https://github.com/rust-lang/rust/issues/110460
+        if 'rust' in compilers and kwargs.get('rust_crate_type', 'rlib') == 'rlib' and any(c in name for c in ['-', ' ']):
```

Comment:
I think the main problem with this approach is that somewhere in `BuildTarget.__init__` it initializes `self.compilers` with only the ones that apply. So we would have to ideally do this afterwards. But then something might've used the wrong name already in the meantime.

Also the defaulting of the crate type is now duplicated with `post_init`.